# Documentation

*   [DeepVariant release notes](https://github.com/google/deepvariant/releases)
*   [Building and testing DeepVariant](deepvariant-build-test.md)
*   [DeepVariant quick start](deepvariant-quick-start.md)
*   [DeepVariant via Docker](deepvariant-docker.md)
*   [DeepVariant whole genome case study](deepvariant-case-study.md)
*   [DeepVariant exome case study](deepvariant-exome-case-study.md)
*   [DeepVariant Genomic VCF (gVCF) support](deepvariant-gvcf-support.md)
*   [DeepVariant usage guide](deepvariant-details.md)
*   [Advanced Case Study: Train a customized SNP and small indel variant caller
    for BGISEQ-500 data](deepvariant-tpu-training-case-study.md)
*   [DeepVariant model training (old)](deepvariant-model-training.md)
*   [DeepVariant training data](deepvariant-details-training-data.md)
*   [Datalab example: visualizing pileup images/tensors](visualizing_examples.ipynb)
*   [Getting Started with GCP](deepvariant-gcp-info.md) (It is not required to
    run DeepVariant on GCP.)

# Other resources

*   [Google Developer Codelab: Variant Calling on a Rice genome with
    DeepVariant](https://codelabs.developers.google.com/codelabs/genomics-deepvariant)
*   [Improve DeepVariant for BGISEQ germline variant
    calling](http://bit.ly/train-deepvariant) |
    [slides](https://github.com/SVAI/RecausalNucleotideNetworks/blob/master/ReCausalNucleotideNetwork.pdf)
